File: 백준/Gold/2467. 용액/용액.py
n = int(input())
value = list(map(int, input().split()))
left, right = 0, n - 1
check, res1, res2 = abs(value[left] + value[right]), value[left], value[right]

while left < right:
    tmp = value[left] + value[right]
    if abs(tmp) < check:
        check = abs(tmp)
        res1 = value[left]
        res2 = value[right]
        if check == 0:
            break
    if tmp > 0:
        right -= 1
    else:
        left += 1
print(res1, res2)

File: 백준/Gold/2467. 용액/용액.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

class Main {

	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		int N = Integer.parseInt(br.readLine());

		long[] nums = new long[N];
		StringTokenizer st = new StringTokenizer(br.readLine());
		for (int i = 0; i < N; i++) {
			nums[i] = Long.parseLong(st.nextToken());
		}

		int left = 0, right = N - 1;
		long value;
		long ansL = -1, ansR = -1;
		long minValue = 2000000000L;
		while (left < right) {
			value = nums[left] + nums[right];
			if (minValue >= Math.abs(value)) {
				minValue = Math.abs(value);
				ansL = nums[left];
				ansR = nums[right];
			}
			if (value > 0) {
				right--;
			} else if (value < 0) {
				left++;
			} else {
				break;
			}
		}
		System.out.println(ansL + " " + ansR);
	}
}